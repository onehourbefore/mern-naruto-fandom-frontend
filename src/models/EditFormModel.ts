

export enum TasksEnum {
    CREATE= 'create',
    UPDATE = 'update',
    IDLE = 'idle'
}

export interface EditFormState {
    postID: string
    task: TasksEnum
}