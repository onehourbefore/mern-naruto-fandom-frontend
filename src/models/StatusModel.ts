export enum StatusModel {
    LOADING = 'loading',
    ERROR = 'error',
    SUCCESS = 'success',
    IDLE = 'idle'
}