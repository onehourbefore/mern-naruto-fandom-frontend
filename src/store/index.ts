import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux';

import posts from './slices/postsSlice';
import filter from './slices/filterSlice';
import onePost from './slices/onePostSlice';
import registration from './slices/registrationSlice';
import authorization from './slices/authorizationSlice';
import experts from './slices/expertSlice';
import comments from './slices/commentSlice';
import editForm from './slices/editFormSlice';

export const rootReducer = combineReducers ({
    posts,
    filter,
    onePost,
    registration,
    authorization,
    experts,
    comments,
    editForm,
});

export const store = configureStore ({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware (),
    devTools: process.env.NODE_ENV !== 'production'
});

export type RootState = ReturnType <typeof store.getState>;
export const useAppDispatch: () => typeof store.dispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;