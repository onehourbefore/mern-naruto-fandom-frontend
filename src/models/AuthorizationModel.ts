import { StatusModel } from "./StatusModel"

export type UserDataType = {
    id: string,
    email: string,
    name: string,
    role: string,
    avatar: string,
    created: number,
    liked: string [] | []
}

export type ResponseType = {
    accessToken: string,
    refreshToken: string,
    user: UserDataType
}

export interface AuthorizationState {
    user: UserDataType,
    status: StatusModel,
    deleteProfileStatus: StatusModel
}