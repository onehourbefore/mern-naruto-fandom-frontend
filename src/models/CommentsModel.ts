import { StatusModel } from "./StatusModel"

export type CommentType = {
    _id: string,
    postID: string,
    date: string,
    name: string,
    email: string,
    body: string,
    avatar: string,
}

export interface CommentsState {
    comments: [] | CommentType [],
    status: StatusModel,
}

export type CommentProps = {
    data: {
        _id: string,
        name: string,
        email: string,
        body: string,
        avatar: string,
        date: string,
    },
    postID: string | undefined,
    commentIndex: number,
}

export type CommentFormProps = {
    setVisible: (arg: boolean) => void,
}

export type CommentFromSidebarProps = {
    comment: CommentType,
}