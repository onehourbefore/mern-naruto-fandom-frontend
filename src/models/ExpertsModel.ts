import { StatusModel } from "./StatusModel"

export type ExpertType = {
    avatar: string,
    email: string,
    name: string,
    created: number
}


export interface ExpertsState {
    experts: [] | ExpertType [],
    status: StatusModel
}