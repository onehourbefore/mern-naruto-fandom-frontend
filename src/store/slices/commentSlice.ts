import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import $api from '../../http';
import { CommentsThunks } from './thunks/CommentsThunks';
import { CommentType, CommentsState } from '../../models/CommentsModel';
import { StatusModel } from '../../models/StatusModel';

const commentsThunks = new CommentsThunks();

const initialState: CommentsState = {
    comments: [],
    status : StatusModel.IDLE,
}

export const fetchComments = createAsyncThunk(
    'fetchComments',
    async(_, { rejectWithValue }) => {
        try {
            return await commentsThunks.fetchCommentsCallback();
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
)

export const commentSlice = createSlice({
    name: 'comments',
    initialState,
    reducers: {
        clearComments: (state) => {
            state.comments = [];
        }
    },
    extraReducers: builder => {
        builder
        .addCase (fetchComments.pending, (state) => {
            state.status = StatusModel.LOADING;
        })
        .addCase (fetchComments.fulfilled, (state, action: PayloadAction<CommentType[]>) => {
            state.comments = action.payload;
            state.status = StatusModel.SUCCESS;
        })
        .addCase (fetchComments.rejected, (state) => {
            state.status = StatusModel.ERROR;
        })
        
        .addDefaultCase (() => {})
    }
})

const { actions, reducer } = commentSlice;
export default reducer;
export const { clearComments } = actions;