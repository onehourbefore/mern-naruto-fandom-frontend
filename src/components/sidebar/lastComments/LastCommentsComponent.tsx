import React from 'react';
import { useAppDispatch, useAppSelector } from '../../../store';
import { getComments, getPostsAll } from '../../../selectors';
import { fetchComments } from '../../../store/slices/commentSlice';
import { CommentType } from '../../../models/CommentsModel';
import { StatusModel } from '../../../models/StatusModel';
import CommentFromSidebar from './CommentFromSidebar';
import Spinner from '../../spinner/Spinner';

import cl from './LastCommentsComponent.module.scss';


const LastCommentsComponent: React.FC = (): JSX.Element => {
    const dispatch = useAppDispatch();
    const { posts } = useAppSelector(getPostsAll);
    const { comments, status } = useAppSelector(getComments);
    const lastComments = showLastComments();

    function showLastComments(): CommentType[] {
        let lastCommentaries: CommentType[] = [];
        if (comments.length <= 3) {
            for (let i = comments.length - 1; i >= 0; i--) {
                lastCommentaries.push(comments[i]);
            }
            return lastCommentaries;
        }
        for (let i = comments.length - 1; i > comments.length - 4; i--) {
            lastCommentaries.push(comments[i]);
        }
        return lastCommentaries;
    };

    React.useEffect(() => {
        if (posts.length) {
            dispatch(fetchComments());
        }
    }, [posts.length]);

    return (
        <div className={cl.root}>
            {status === StatusModel.LOADING && 
                <Spinner mt='10px' />}

            {lastComments.length !== 0 && 
                <h3 className={cl.root__title}>
                    ПОСЛЕДНИЕ КОММЕНТАРИИ
                </h3>}
                
            {lastComments.map(comment => 
                <CommentFromSidebar 
                    key={comment._id}
                    comment={comment}
                />)}
        </div>
    )
};

export default LastCommentsComponent;