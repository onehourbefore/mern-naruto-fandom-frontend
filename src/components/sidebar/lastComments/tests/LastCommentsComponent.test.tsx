import { screen, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { StatusModel } from '../../../../models/StatusModel';
import * as reduxHooks from '../../../../store';
import * as actions from '../../../../store/slices/commentSlice';
import LastCommentsComponent from '../LastCommentsComponent';
import { returnedStateLastCommentsTest as state } from './dataForTests';


const mockedUseAppDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockedUseAppSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const mockedFetchCommentsAction = jest.spyOn(actions, 'fetchComments');


describe('LastCommentsComponent', () => {
    it('fetch status loading', async () => {
        // console.log(String(new Date()));
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue({
            ...state,
            posts: [],
            comments: [],
            status: StatusModel.LOADING,
        });

        const component = render(
            <MemoryRouter>
                <LastCommentsComponent />
            </MemoryRouter>
        );
        const spinner = await screen.findByTestId('spinner');
        expect(spinner).toBeInTheDocument();
        expect(mockedFetchCommentsAction).toHaveBeenCalledTimes(0);
        expect(component).toMatchSnapshot();
    });

    it('fetch status success', async () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue(state);

        const component = render(
            <MemoryRouter>
                <LastCommentsComponent />
            </MemoryRouter>
        );
        expect(screen.getByText(/последние комментарии/i)).toBeInTheDocument();

        expect(screen.getByText('John')).toBeInTheDocument();
        expect(screen.getByText('john@gmail.com')).toBeInTheDocument();
        expect(screen.getByText('Oops')).toBeInTheDocument();

        expect(screen.getByText('Andy')).toBeInTheDocument();
        expect(screen.getByText('andy@gmail.com')).toBeInTheDocument();
        expect(screen.getByText('I like football')).toBeInTheDocument();
        
        expect(screen.getByText('Rob')).toBeInTheDocument();
        expect(screen.getByText('rob@gmail.com')).toBeInTheDocument();
        expect(screen.getByText('Welcome!')).toBeInTheDocument();

        expect(mockedFetchCommentsAction).toHaveBeenCalledTimes(1);
        expect(component).toMatchSnapshot();
    });
});