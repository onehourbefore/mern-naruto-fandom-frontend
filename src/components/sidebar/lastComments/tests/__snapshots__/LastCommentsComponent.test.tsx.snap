// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastCommentsComponent fetch status loading 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div
          class="spinner"
          data-testid="spinner"
          style="margin-top: 10px;"
        >
          <div
            class="ldsRing"
          >
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div
        class="spinner"
        data-testid="spinner"
        style="margin-top: 10px;"
      >
        <div
          class="ldsRing"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LastCommentsComponent fetch status success 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <h3
          class="root__title"
        >
          ПОСЛЕДНИЕ КОММЕНТАРИИ
        </h3>
        <div
          class="root__comment"
          data-testid="commentFromSidebar-test"
        >
          <div
            class="root__comment_avatar"
          >
            <img
              alt="Аватар"
              class="root__comment_avatarIcon"
              src="http://localhost:5000/avatars/johnAvatar.jpg"
            />
          </div>
          <div
            class="root__comment_content"
          >
            <div
              class="root__comment_header"
            >
              <div>
                <h3
                  class="root__comment_header_title"
                >
                  John
                </h3>
                <div
                  class="root__comment_header_email"
                >
                  john@gmail.com
                </div>
              </div>
              <div>
                <div
                  class="root__comment_header_date"
                >
                  <div>
                    13 Nov 2022
                  </div>
                  <div>
                    07:43
                  </div>
                </div>
              </div>
            </div>
            <div
              class="root__comment_body"
            >
              Oops
            </div>
          </div>
        </div>
        <div
          class="root__comment"
          data-testid="commentFromSidebar-test"
        >
          <div
            class="root__comment_avatar"
          >
            <img
              alt="Аватар"
              class="root__comment_avatarIcon"
              src="http://localhost:5000/avatars/andyAvatar.jpg"
            />
          </div>
          <div
            class="root__comment_content"
          >
            <div
              class="root__comment_header"
            >
              <div>
                <h3
                  class="root__comment_header_title"
                >
                  Andy
                </h3>
                <div
                  class="root__comment_header_email"
                >
                  andy@gmail.com
                </div>
              </div>
              <div>
                <div
                  class="root__comment_header_date"
                >
                  <div>
                    12 Nov 2022
                  </div>
                  <div>
                    08:10
                  </div>
                </div>
              </div>
            </div>
            <div
              class="root__comment_body"
            >
              I like football
            </div>
          </div>
        </div>
        <div
          class="root__comment"
          data-testid="commentFromSidebar-test"
        >
          <div
            class="root__comment_avatar"
          >
            <img
              alt="Аватар"
              class="root__comment_avatarIcon"
              src="http://localhost:5000/avatars/robAvatar.jpg"
            />
          </div>
          <div
            class="root__comment_content"
          >
            <div
              class="root__comment_header"
            >
              <div>
                <h3
                  class="root__comment_header_title"
                >
                  Rob
                </h3>
                <div
                  class="root__comment_header_email"
                >
                  rob@gmail.com
                </div>
              </div>
              <div>
                <div
                  class="root__comment_header_date"
                >
                  <div>
                    12 Nov 2022
                  </div>
                  <div>
                    08:08
                  </div>
                </div>
              </div>
            </div>
            <div
              class="root__comment_body"
            >
              Welcome!
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <h3
        class="root__title"
      >
        ПОСЛЕДНИЕ КОММЕНТАРИИ
      </h3>
      <div
        class="root__comment"
        data-testid="commentFromSidebar-test"
      >
        <div
          class="root__comment_avatar"
        >
          <img
            alt="Аватар"
            class="root__comment_avatarIcon"
            src="http://localhost:5000/avatars/johnAvatar.jpg"
          />
        </div>
        <div
          class="root__comment_content"
        >
          <div
            class="root__comment_header"
          >
            <div>
              <h3
                class="root__comment_header_title"
              >
                John
              </h3>
              <div
                class="root__comment_header_email"
              >
                john@gmail.com
              </div>
            </div>
            <div>
              <div
                class="root__comment_header_date"
              >
                <div>
                  13 Nov 2022
                </div>
                <div>
                  07:43
                </div>
              </div>
            </div>
          </div>
          <div
            class="root__comment_body"
          >
            Oops
          </div>
        </div>
      </div>
      <div
        class="root__comment"
        data-testid="commentFromSidebar-test"
      >
        <div
          class="root__comment_avatar"
        >
          <img
            alt="Аватар"
            class="root__comment_avatarIcon"
            src="http://localhost:5000/avatars/andyAvatar.jpg"
          />
        </div>
        <div
          class="root__comment_content"
        >
          <div
            class="root__comment_header"
          >
            <div>
              <h3
                class="root__comment_header_title"
              >
                Andy
              </h3>
              <div
                class="root__comment_header_email"
              >
                andy@gmail.com
              </div>
            </div>
            <div>
              <div
                class="root__comment_header_date"
              >
                <div>
                  12 Nov 2022
                </div>
                <div>
                  08:10
                </div>
              </div>
            </div>
          </div>
          <div
            class="root__comment_body"
          >
            I like football
          </div>
        </div>
      </div>
      <div
        class="root__comment"
        data-testid="commentFromSidebar-test"
      >
        <div
          class="root__comment_avatar"
        >
          <img
            alt="Аватар"
            class="root__comment_avatarIcon"
            src="http://localhost:5000/avatars/robAvatar.jpg"
          />
        </div>
        <div
          class="root__comment_content"
        >
          <div
            class="root__comment_header"
          >
            <div>
              <h3
                class="root__comment_header_title"
              >
                Rob
              </h3>
              <div
                class="root__comment_header_email"
              >
                rob@gmail.com
              </div>
            </div>
            <div>
              <div
                class="root__comment_header_date"
              >
                <div>
                  12 Nov 2022
                </div>
                <div>
                  08:08
                </div>
              </div>
            </div>
          </div>
          <div
            class="root__comment_body"
          >
            Welcome!
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
